### PR TITLE
In glib_object_wrapper!, use generic Class type instead defining a struct

### PR DESCRIPTION
**Pull request by ids1024**
Update of https://github.com/gtk-rs/glib/pull/720 with corresponding changes to other crates.

This makes the `$rust_class_name` argument in `glib_wrapper!` optional. If specified, it defines a `type` alias for the class. At least optionally allowing it is required to avoid breaking lots of gir generated code, but perhaps it can be removed after a change to gir.

This removes the `ObjectType::RustClassType` associated type, since it can't be enforced/assumed by the type system that that is an instance of this generic struct.

The `IsClassFor` trait is also removed, superseded by this generic type.

This also changes the `IsSubclassable` trait to be implemented on the object struct rather than the class struct, since there otherwise would be foreign trait on foreign type errors.

### CODE REVIEW

**Comment by sdroege** [comment]
Can you rebase on top of latest master? Then we get CI back :)

File: glib/src/object.rs
Comment (sdroege) [inline glib/src/object.rs:2518]: Instead of that, can we get `AsRef<T>` and `AsMut<T>`? If not, naming them `as_ref()` and `as_mut()` would seem nicer nonetheless

File: glib/src/object.rs
Comment (sdroege) [inline glib/src/object.rs:56]: So far we didn't require there to be a wrapper type for `ObjectSubclass` impls... maybe that's something we can also change in the future as it seems like it would make things easier, or not?

Also it would allow returning the type itself instead of the parent type from `ObjectSubclass::get_instance()` and similar weirdnesses that we have right now.

What do you think @ids1024? We can do that separate from this PR though, no need to block on this here.

File: gtk/src/subclass/application.rs
Comment (sdroege) [inline gtk/src/subclass/application.rs:52]: I hope we get custom self types at some point, `self: &mut Class<Self>` (compare `Future::poll`) would look nicer :)

**Review by sdroege — COMMENTED** [review]
Looks good to me otherwise. Thanks a lot!

File: gtk/src/subclass/application.rs
Comment (ids1024) [reply gtk/src/subclass/application.rs:52]: The unstable arbitrary self types feature in Rust, as implemented, only works when the type implements `Deref` to `self`. So assuming that's how a stable feature would work, I don't think it would help here. Unfortunately.

File: glib/src/object.rs
Comment (ids1024) [reply glib/src/object.rs:2518]: Good idea. That would probably be better.

File: glib/src/object.rs
Comment (ids1024) [reply glib/src/object.rs:56]: That could be useful. Similarly to `get_instance`, all of the `*Impl` traits have methods receiving the parent type, which a lot of implementations will immediately want to downcast.

I've been thinking of that recently. I wonder if subclassing should requiring defining a `glib_wrapper!` and specifying that as an associated type in `ObjectSubclass`. If that doesn't lead to recursive type errors, and doesn't make basic subclassing too hard, it could improve some of these things.

But yeah, that's separate and doesn't need to block this PR.

File: glib/src/object.rs
Comment (ids1024) [reply glib/src/object.rs:2518]: Though that does have the downside that it would probably require callers to specify the type, since it can't be inferred that `as_ref()` can only work with one type.

File: gtk/src/subclass/application.rs
Comment (sdroege) [reply gtk/src/subclass/application.rs:52]: Ah too bad. We can live with that, I guess

File: glib/src/object.rs
Comment (sdroege) [reply glib/src/object.rs:56]: Yeah passing it to the virtual methods was also what I was thinking of, that should be possible then. Can you create an issue for this?

File: glib/src/object.rs
Comment (ids1024) [reply glib/src/object.rs]: But I think that's fine, since it may be more readable if code using this (which would probably mostly be implementations of `IsSubclassable`) is explicit about the type anyway.

I've changed this.

File: glib/src/object.rs
Comment (ids1024) [reply glib/src/object.rs:56]: Opened https://github.com/gtk-rs/gtk-rs/issues/14

Comment (sdroege) [reply glib/src/object.rs:56]: Thanks!

**Comment by sdroege** [comment]
I'll merge this later today or tomorrow once I've updated `gstreamer-rs` accordingly. Don't want it to fail once this is merged :)

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

File: glib/src/wrapper.rs
Comment (sdroege) [inline glib/src/wrapper.rs:338]: @ids1024 Can you also do the `gir` change that makes it use the new variants without the explicit class name?

File: gtk/src/subclass/cell_renderer_text.rs
Comment (sdroege) [inline gtk/src/subclass/cell_renderer_text.rs:44]:
```suggestion
            let klass = &mut *(class.as_mut() as *mut _ as *mut gtk_sys::GtkCellRendererTextClass);
```

Might be nicer but doesn't really matter

Comment (sdroege) [reply gtk/src/subclass/cell_renderer_text.rs:44]: Especially no wildcard cast anymore :)

Comment (GuillaumeGomez) [reply gtk/src/subclass/cell_renderer_text.rs:44]: Oh I like it!

Comment (sdroege) [reply gtk/src/subclass/cell_renderer_text.rs:44]: `sed -E -i 's;klass as \*mut _ ;klass.as_mut() ;' */src/subclass/*.rs`

Changes this automatically btw

**Comment by sdroege** [comment]
> I'll merge this later today or tomorrow once I've updated `gstreamer-rs` accordingly. Don't want it to fail once this is merged :)

That's waiting here now https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/merge_requests/621

File: gtk/src/subclass/cell_renderer_text.rs
Comment (sdroege) [reply gtk/src/subclass/cell_renderer_text.rs]: Updated the commit so once the CI is through we can merge it here.

Comment (ids1024) [reply gtk/src/subclass/cell_renderer_text.rs]: Actually, that should be equivalent to just `let klass = class.as_mut();`. Which is exactly what I was thinking with the `AsMut` implementation.

https://github.com/gtk-rs/gtk-rs/pull/16

Comment (sdroege) [reply gtk/src/subclass/cell_renderer_text.rs]: Oh even simpler. Thanks!

File: glib/src/wrapper.rs
Comment (ids1024) [reply glib/src/wrapper.rs:338]: It doesn't look complicated to update `gir`, but I'm seeing issues running `generator.py` with `gir` master similar to those mentioned in https://github.com/gtk-rs/gtk-rs/pull/11

Comment (sdroege) [reply glib/src/wrapper.rs:338]: Indeed, something's broken currently. Can you create an issue for gir otherwise?

**Pull request merged**
